File: backend/app/schemas/placement.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class PlacementChoiceOut(BaseModel):
    key: str
    text: str


class PlacementQuestionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    seq: int
    q_type: str
    stem: str
    choices: list[PlacementChoiceOut] = Field(default_factory=list)
    answer_key: str | None = None
    points: int = 1


class PlacementPaperSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    subject_code: str
    status: str
    title: str
    created_at: datetime


class PlacementPaperDetail(PlacementPaperSummary):
    questions: list[PlacementQuestionOut]
    gen_job_id: uuid.UUID | None = None


class PlacementSubjectStatus(BaseModel):
    subject_code: str
    status: str
    paper_id: uuid.UUID | None = None


class PlacementStartIn(BaseModel):
    subject_code: str | None = None


class PlacementStartOut(BaseModel):
    subjects: list[PlacementSubjectStatus]
    gen_job_id: uuid.UUID | None = None


class PlacementAnswerIn(BaseModel):
    question_id: uuid.UUID
    content: str = Field(min_length=1, max_length=10000)


class PlacementSubmitIn(BaseModel):
    answers: list[PlacementAnswerIn]


class PlacementSubmitOut(BaseModel):
    paper_id: uuid.UUID
    total_score: int
    mastery_json: dict
    roadmap_job_id: uuid.UUID | None = None
    all_placement_complete: bool = False

File: frontend/src/api/placement.ts
import { api } from "./client";

export type PlacementStartOut = {
  subjects: { subject_code: string; status: string; paper_id: string | null }[];
  gen_job_id?: string | null;
};

export type PlacementPaperSummary = {
  id: string;
  subject_code: string;
  status: string;
  title: string;
  created_at: string;
};

export type PlacementChoiceOut = { key: string; text: string };

export type PlacementQuestionOut = {
  id: string;
  seq: number;
  q_type: string;
  stem: string;
  choices: PlacementChoiceOut[];
  answer_key: string | null;
};

export type PlacementPaperDetail = PlacementPaperSummary & {
  questions: PlacementQuestionOut[];
};

export type PlacementSubmitIn = {
  answers: { question_id: string; content: string }[];
};

export type PlacementSubmitOut = {
  paper_id: string;
  total_score: number;
  mastery_json: Record<string, number>;
};

export function startPlacement(body?: { subject_code?: string }) {
  return api<PlacementStartOut>("/student/placement/start", {
    method: "POST",
    body: JSON.stringify(body ?? {}),
  });
}

export function listPlacementPapers() {
  return api<PlacementPaperSummary[]>("/student/placement");
}

export function getPlacementPaper(paperId: string) {
  return api<PlacementPaperDetail>(`/student/placement/${paperId}`);
}

export function submitPlacement(paperId: string, payload: PlacementSubmitIn) {
  return api<PlacementSubmitOut>(`/student/placement/${paperId}/submit`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

